beps/lib-kafka/1374_timestamp_based_seek.md

```diff
@@ -0,0 +1,150 @@
+# Timestamp-based seeking support and retrieve metadata when sending message in ballerinax/kafka
+
+- Authors
+  - Thisaru Guruge
+- Reviewed by
+  - Danesh Kuruppu
+- Created date 
+  - 2025-07-08 
+- Updated date
+  - 2025-07-08
+- Issue
+  - [1374](https://github.com/ballerina-platform/ballerina-spec/issues/1374)
+- State
+  - Submitted
+
+## Summary
+
+The current Ballerina Kafka connector lacks some features which helps the users to improve their performance, efficiency, and ease of use. This aim of this proposal is to introduce new APIs to improve the performance and the user experience.
+
+## Motivation
+
+In real-world scenarios, having the ability to retrieve the metadata when sending a message to a Kafka broker, or seeking a consumer for a desired timestamp, are a few must-have features. This provides better troubleshooting capabilities, recoverability, performance improvement, and better user experience.
+
+## Goals
+
+* Provide the Kafka consumer an easy way to seek to a given timestamp
+* Provide an API to retrieve metadata when sending Kafka records 
+
+## Design
+
+The following APIs are proposed to solve the aforementioned problems.
+
+### Producer Send API
+The current producer API has the following definition:
+
+```ballerina
+isolated remote function send(AnydataProducerRecord producerRecord) returns Error?
+```
+
+This returns an error if the sending is failed, but does not return any metadata when the record is successfully sent. This metadata is useful for troubleshooting and observability purposes.
+
+#### The RecordMetadata Type
+
+A new record type will be introduced to represent metadata that is retrieved while sending a message. This type directly corresponds to the RecordMetadata class in Apache Kafka Java SDK.
+
+```ballerina
+# Represents metadata of a Kafka record.
+public type RecordMetadata record {|
+# Indicates whether the record metadata has an offset.
+boolean hasOffset;
+# The offset of the record in the topic partition
```

Comment:
Can't we combine these to and provide an optional field as `offset` ?

Reply:
Good catch! Fixed with 43b2ae305a3f2839aa891035a8c37ed50ba54a34